### DISCUSSION TITLE
Rich Notification Service Extensions

### DISCUSSION THREAD

**Discussion by mvolonnino** [expo/expo#25162] [Proposals and Ideas]
I can not seem to figure out how to add Rich Notification Service Extensions for iOS push notifications. I utilize Braze and FCM for push notification handling. I have set up Rich Notification services before for Braze on a bare RN App, but my work is re platforming to use the Expo tooling which is fantastic.

Is there ways to add this now in Expo?

I've read this which I do not fully understand:
https://docs.expo.dev/build-reference/app-extensions/#managed-projects-experimental-support

And also have read this, albeit not fully understanding it as well:
https://docs.expo.dev/config-plugins/plugins-and-mods/

I figure a combination of something like that could get something to work, but would need some clarification / help in understanding how to create expo plugins + if its even possible with the Expo framework of getting rich notification services to work on all incoming push notifications if they have an image attached to them!

Thanks a lot to anyone who is able to shed some light

**Comment by angelo-hub** [comment]
i had the same question, my company is going about procuring braze and i found this cool extension that does it for oneSignal; i may port it to add braze compatibility since their rich notifications act similar 

https://github.com/OneSignal/onesignal-expo-plugin

**Reply by mvolonnino** [reply]
That would be huge! Have not been able to put anymore time into this yet, but was hoping that something like this would get integrated into the Braze config plugin as well 🙌

**Comment by julianD77** [comment]
AFAIK you should be able to use [Expo Notifications](https://docs.expo.dev/versions/latest/sdk/notifications/) to achieve this. Is that not the case? 

We are starting the process of migrating our App to Expo and were planning on using `expo-notifications` with our existing FCM notifications. 

I know it refers to Firebase Cloud Messaging (FCM) rather than Braze, but does the following doc help you?

https://docs.expo.dev/push-notifications/sending-notifications-custom/

**Comment by nikwebr** [comment]
Hey, in case anybody is still interested, I modified the [onesignal-expo-plugin](https://github.com/OneSignal/onesignal-expo-plugin) to a config plugin that simply takes a notification service extension file and copies it in the app binaries. 
The plugin is [here](https://github.com/nikwebr/expo-notification-service-extension-plugin) and some additional documentation can be found [here](https://checkokay.com/blog/nse).

**Reply by sahiljaidka224** [reply]
this is amazing, thanks

**Reply by sahiljaidka224** [reply]
@nikwebr I can increase the count using your plugin but how can I reset?

This is what I added in NotificationService.m file, from expo I can doing Notifications.setBadgeCountAsync(0) which was working fine before but now it clears, after a new notification is received it starts the counter from like 10 which was prev number

```
NSUserDefaults *defaults = [[NSUserDefaults alloc] initWithSuiteName:@"com.galaxies.swiper"];

    NSInteger count = [[defaults valueForKey:@"count"] integerValue];
    if (self.bestAttemptContent != nil) {
        NSInteger receivedBadge = [self.bestAttemptContent.badge integerValue];
        
        if (count == 0) {
            if (receivedBadge > 0) {
                count = receivedBadge;
            }
        } else {
            count += 1;
        }
        
        self.bestAttemptContent.badge = @(count);
        [defaults setInteger:count forKey:@"count"];
        
        self.contentHandler(self.bestAttemptContent);
    } else {
        self.contentHandler(self.bestAttemptContent);
    }
```

**Reply by nikwebr** [reply]
Using your above code, you need to reset the NSUserDefaults for `count` to zero on all places where you do `Notifications.setBadgeCountAsync(0)`

One question, why do you use this line
`NSInteger receivedBadge = [self.bestAttemptContent.badge integerValue];` ?

**Reply by sahiljaidka224** [reply]
Can you recommend any library I can use to reset `NSUserDefaults`?

for receivedBadge, you're right I probably don't need it

**Reply by nikwebr** [reply]
I am using the [@alevy97/react-native-userdefaults](https://www.npmjs.com/package/@alevy97/react-native-userdefaults?activeTab=readme) package.

It lets you modify shared user defaults in an app group by initializing it this way
`const groupDefaults = new UserDefaults("com.galaxies.swiper");`

**Reply by sahiljaidka224** [reply]
@nikwebr Are you using Expo? doesn't seem to work for me in Expo, having TS errors

**Reply by sahiljaidka224** [reply]
import UserDefaults from "@alevy97/react-native-userdefaults/src/ReactNativeUserDefaults.ios";

I've to import like this to make it work.

**Reply by sahiljaidka224** [reply]
and it doesn't even work, like it's not updating number count back to 0

**Comment by pdecarcer** [comment]
If you are using Firebase package and not expo-notifications you can use our extension.  [https://www.npmjs.com/package/@abstract-cl/expo-firebase-extension-plugin](url)

**Reply by marcel-happyfloat** [reply]
is this repo private, I can't find it?

**Reply by joaqo** [reply]
Same here, can't find it.

**Comment by marcel-happyfloat** [comment]
This snippet worked for me, credits to the Author of it!
https://gist.github.com/bensaine/4feedd8d6799454640b5e1f0ffd267df
